### PR TITLE
add /latest/dynamic/instance-identity/document

### PR DESCRIPTION
**Pull request by bobbyc**
*Issue #, if available:*

#70 

*Description of changes:*

Add /latest/dynamic/instance-identity/document for simulate instance-identity. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by bwagner5** [comment]
Hey @bobbyc thanks for the PR!!

There were a few issues found by the travis CI run.  Looks like most of the problems can be fixed if you run `make fmt`. 

```
❌ goimports found formatting issues in go source files. See below for the files with problems.
Grade: A+ (92.7%)
Files: 34
Issues: 6
gofmt: 82%
	pkg/config/types.go
		Line 1: warning: file is not gofmted with -s (gofmt)
	pkg/mock/dynamic/dynamic.go
		Line 1: warning: file is not gofmted with -s (gofmt)
	pkg/mock/dynamic/types/types.go
		Line 1: warning: file is not gofmted with -s (gofmt)
	pkg/mock/listmocks/listmocks.go
		Line 1: warning: file is not gofmted with -s (gofmt)
	pkg/mock/static/static.go
		Line 1: warning: file is not gofmted with -s (gofmt)
	pkg/config/dynamic.go
		Line 1: warning: file is not gofmted with -s (gofmt)
go_vet: 100%
gocyclo: 100%
golint: 91%
	pkg/mock/dynamic/dynamic.go
		Line 31: warning: exported var ServicePath should have comment or be unexported (golint)
	pkg/mock/static/static.go
		Line 31: warning: exported var ServicePath should have comment or be unexported (golint)
	pkg/config/types.go
		Line 56: warning: exported type Dynamic should have comment or be unexported (golint)
		Line 171: warning: exported type DynamicPaths should have comment or be unexported (golint)
		Line 175: warning: comment on exported type DynamicValues should be of the form "DynamicValues ..." (with optional leading article) (golint)
ineffassign: 100%
license: 100%
misspell: 100%
❌ Test failed to meet go-report-card threshold score of: 100
```

Looks like some of the e2e tests also failed.  I haven't looked into why yet, but I do see this in the log, not sure if it's the cause or not: 

```
There was an issue registering path  with mdValue: {              }
```

**Comment by bobbyc** [comment]
Hi @bwagner5, 

all set. Since I don't have enough time to work on this, please help me to write test cases from your team. Thanks.

File: pkg/mock/static/static.go [hunk 1]
```diff
@@ -27,6 +27,11 @@ var (
 	supportedPaths   = make(map[string]interface{})
 	response         interface{}
 	jsonTextResponse = map[string]bool{"/latest/meta-data/elastic-inference/associations/eia-bfa21c7904f64a82a21b9f4540169ce1": true}
+
+	// ServicePath defines the static service path
+	ServicePath = "/latest/meta-data"
+	// ServicePath2 defines the static service path with slash
+	ServicePath2 = ServicePath + "/"
```
Comment (brycahta) [inline pkg/mock/static/static.go:34]: thanks👍🏼

File: pkg/mock/dynamic/dynamic.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+	cfg "github.com/aws/amazon-ec2-metadata-mock/pkg/config"
+	"github.com/aws/amazon-ec2-metadata-mock/pkg/mock/imdsv2"
+	"github.com/aws/amazon-ec2-metadata-mock/pkg/server"
+)
+
+var (
+	supportedPaths   = make(map[string]interface{})
+	response         interface{}
+	jsonTextResponse = map[string]bool{"": true}
```
Comment (brycahta) [inline pkg/mock/dynamic/dynamic.go:29]: according to [aws docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html) `instance-identity/document` returns a "**JSON** containing instance attributes,..."

`jsonTextResponse` and accompanying logic can be removed if the response is truly a JSON

File: pkg/config/dynamic.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	// Unmarshal to map to preserve keys for Paths and Values
+	var defaultsMap map[string]interface{}
+	json.Unmarshal(jsonWithDefaults, &defaultsMap)
+
+	dyPaths := defaultsMap["dynamic"].(map[string]interface{})["paths"].(map[string]interface{})
+	dyValues := defaultsMap["dynamic"].(map[string]interface{})["values"].(map[string]interface{})
+
+	for k, v := range dyPaths {
+		newKey := dyPathsCfgPrefix + k
+		// ex: "dynamic.paths.ami-id": "/latest/meta-data/ami-id"
```
Comment (brycahta) [inline pkg/config/dynamic.go:73]: can you update the comment to reflect an accurate example of dynamic path:value or omit it completely

File: pkg/config/dynamic.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package config
+
+import (
+	"encoding/json"
+
+	// "github.com/aws/amazon-ec2-metadata-mock/pkg/mock/dynamic/types"
```
Comment (brycahta) [inline pkg/config/dynamic.go:19]: remove unused imports

**Comment by brycahta** [comment]
Hi @bobbyc , appreciate the contribution. We have a policy on our repos that any new feature additions require accompanying tests; therefore, we won't be able to merge these changes until the tests are added.

Our team will work on updating the PR template to state this more clearly.

File: pkg/config/dynamic.go (hunk 4 above)
Comment (bobbyc) [reply pkg/config/dynamic.go]: removed

File: pkg/config/dynamic.go (hunk 3 above)
Comment (bobbyc) [reply pkg/config/dynamic.go]: updated

File: pkg/mock/dynamic/dynamic.go (hunk 2 above)
Comment (bobbyc) [reply pkg/mock/dynamic/dynamic.go]: according to the template Handler() code, I only remove the map. And leave the jsonTextResponse variable for future usage.

**Comment by bobbyc** [comment]
@brycahta I update a simple e2e-test for root paths and healthy check. Not very sure this is all you need. Let me know. Thanks.

**Comment by brycahta** [comment]
@bobbyc , in addition to the health check + verifying paths via golden file, could you also test the following:
* the default values return as expected 
* values are overridden as expected when using a config file

**static-test** does both of these things already if you need guidance. You can either add another test function to this file or create a new **dynamic-test** file-- up to you

**Comment by bobbyc** [comment]
**Testing**:

- Service path and Healthy check
- Default and overridden values check based on @brycahta code review and suggestion

**Pull request merged**